### PR TITLE
Rsh/snippet generation java kiota

### LINKED ISSUE #1026: Java Snippet Generation - OpenAPI
Refactor Java Snippet Generation To Use Code [SnippetCodeGraph](https://github.com/microsoftgraph/microsoft-graph-devx-api/blob/51342d36b6dbd1faf7091a6923d1196b6b4ddd3f/CodeSnippetsReflection.OpenAPI/ModelGraph/SnippetCodeGraph.cs) Class

### PR DESCRIPTION
**Pull request by ramsessanchez**
Fixes #1026 
Adds snippet generation for the Kiota based Java-SDK using OpenApi

### CODE REVIEW

File: CodeSnippetsReflection.OpenAPI/LanguageGenerators/JavaGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+                case PropertyType.Duration:
+                    return "PeriodAndDuration";
+                case PropertyType.Binary:
+                    return "ByteArrayInputStream";
+                default:
+                    return ReplaceIfReservedName(typeString);
+            }
+        }
+
+        private static string GetPropertyObjectName(CodeProperty codeProperty, string apiVersion)
```
Comment (andrueastman) [inline CodeSnippetsReflection.OpenAPI/LanguageGenerators/JavaGenerator.cs:434]:
```suggestion
        private static string GetPropertyObjectName(CodeProperty codeProperty)
```
unused parameter

File: CodeSnippetsReflection.OpenAPI/LanguageGenerators/JavaGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+                    string parameterDeclarations = string.Join(", ", codeGraph.PathParameters
+                                                            .OrderBy(static parameter => parameter.Name, StringComparer.OrdinalIgnoreCase)
+                                                            .Select(parameter => GetPathParameterDeclaration(parameter, codeGraph.ApiVersion)));
+                    return $"{functionName.ToFirstCharacterLowerCase()}{parameters}({parameterDeclarations}).";
+                }
+                if (x.Segment.IsFunction())
+                {
+                    return x.Segment.Split('.')
+                              .Select(s => s.ToFirstCharacterLowerCase())
+                              .Aggregate((a, b) => $"{a}{b}").ToFirstCharacterUpperCase()+"().";
```
Comment (andrueastman) [inline CodeSnippetsReflection.OpenAPI/LanguageGenerators/JavaGenerator.cs:227]:
```suggestion
                              .Select(static s => s.ToFirstCharacterLowerCase())
                              .Aggregate(static (a, b) => $"{a}{b}").ToFirstCharacterUpperCase()+"().";
```

File: CodeSnippetsReflection.OpenAPI/LanguageGenerators/JavaGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+                }
+                return x.Segment.ToFirstCharacterLowerCase()+"().";
+            })
+                        .Aggregate(new List<string>(), (current, next) =>
```
Comment (andrueastman) [inline CodeSnippetsReflection.OpenAPI/LanguageGenerators/JavaGenerator.cs:231]:
```suggestion
                        .Aggregate(new List<string>(), static (current, next) =>
```

**Comment by andrueastman** [comment]
Thanks for this @ramsessanchez. This looks great. Any chance you can checkout the sonarcloud warnings as well?

**Comment by ramsessanchez** [comment]
There is one remaining code smell regarding the use of Last(), the smell suggests that using count-1 would be more efficient, however split returns an array so the smell actually is not helpful.

**Comment by ramsessanchez** [comment]
@andrueastman I believe that we need to add the namespace change to dotNet snippet generation as well. While we can still use .me() the reference to the namespace .me was removed when we removed it during generation thus making some snippets uncompilable.

**Review by andrueastman — APPROVED** [review]
> @andrueastman I believe that we need to add the namespace change to dotNet snippet generation as well. While we can still use .me() the reference to the namespace .me was removed when we removed it during generation thus making some snippets uncompilable.

Thanks for pointing this out @ramsessanchez, I've updated the issue description at https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/1874 to capture this when this is implemented in the next breaking change for dotnet

**Review by andrueastman — DISMISSED** [review]

**Comment by ramsessanchez** [comment]
Ready to merge, will continue fixes to snippet generation on later PR's

File: GraphWebApi/Controllers/SnippetsController.cs [hunk 4]
```diff
@@ -74,7 +74,7 @@ public async Task<IActionResult> PostAsync(string lang = "c#", string generation
         {
             // Default to openapi generation if supported and not explicitly requested for.
             if (string.IsNullOrEmpty(generation))
-                generation = OpenApiSnippetsGenerator.SupportedLanguages.Contains(lang) ? "openapi" : "odata";
+                generation = (lang != "java" && OpenApiSnippetsGenerator.SupportedLanguages.Contains(lang)) ? "openapi" : "odata";
```
Comment (andrueastman) [inline GraphWebApi/Controllers/SnippetsController.cs:77]:
```suggestion
                generation = (!"java" .Equals(lang,StringComparison.OrdinalIgnoreCase)&& OpenApiSnippetsGenerator.SupportedLanguages.Contains(lang)) ? "openapi" : "odata";
```
The dictionary is case insensitive as well so we need to guard against that https://github.com/microsoftgraph/microsoft-graph-devx-api/blob/3c66e2491cc5c32dce561ab24c786b4d06223847/CodeSnippetsReflection.OpenAPI/OpenAPISnippetsGenerator.cs#L77
Same for the other entry point as well.

File: CodeSnippetsReflection.App/Program.cs [hunk 5]
```diff
@@ -96,7 +96,10 @@ static void Main(string[] args)
             var snippetGenerators = new ConcurrentDictionary<string, ISnippetsGenerator>();
             Parallel.ForEach(supportedLanguages, language =>
             {
-                generation = OpenApiSnippetsGenerator.SupportedLanguages.Contains(language) ? "openapi" : originalGeneration;
+                //Generation will still be originalGeneration if language is java since it is not stable
+                //Remove the condition when java is stable
+                generation = (language != "java" && OpenApiSnippetsGenerator.SupportedLanguages.Contains(language)) ? "openapi" : originalGeneration;
```
Comment (andrueastman) [inline CodeSnippetsReflection.App/Program.cs:101]: check for casing

File: CodeSnippetsReflection.OpenAPI/LanguageGenerators/JavaGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+            "try",
+            "void",
+            "volatile",
+            "while"
+        };
+
+        public string GenerateCodeSnippet(SnippetModel snippetModel)
+        {
+            var indentManager = new IndentManager();
+            var codeGraph = new SnippetCodeGraph(snippetModel);
```
Comment (zengin) [inline CodeSnippetsReflection.OpenAPI/LanguageGenerators/JavaGenerator.cs:85]: I am glad that @rkodev's investments are paying dividends here :)

File: CodeSnippetsReflection.OpenAPI/LanguageGenerators/JavaGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+            var responseAssignment = codeGraph.HasReturnedBody() ? $"{codeGraph.ResponseSchema.Reference?.Id.Replace($"{DefaultNamespace}.", string.Empty).ToPascalCase() ?? "var"} result = " : string.Empty;
+            var methodName = codeGraph.HttpMethod.Method.ToLower();
+
+            string requestPayloadParameterName = default;
+            if (codeGraph.HasBody())
+            {
+                if (codeGraph.Body.PropertyType == PropertyType.Binary)
+                    requestPayloadParameterName = "stream";
+                else
+                    requestPayloadParameterName = GetPropertyObjectName(codeGraph.Body).ToFirstCharacterLowerCase();
```
Comment (zengin) [inline CodeSnippetsReflection.OpenAPI/LanguageGenerators/JavaGenerator.cs:105]: nit: easier to read with a switch expression.
```suggestion
                requestPayloadParameterName = codeGraph.Body.PropertyType switch
                {
                    PropertyType.Binary => "stream",
                    _ => GetPropertyObjectName(codeGraph.Body).ToFirstCharacterLowerCase()
                };
```

**Review by zengin — CHANGES_REQUESTED** [review]

**Review by zengin — APPROVED** [review]

**Pull request merged**